js/src/webrtc.js

```diff
@@ -56,6 +57,115 @@ var MediaStreamView = widgets.DOMWidgetView.extend({
+
+    initialize: function() {
+        WidgetStreamModel.__super__.initialize.apply(this, arguments);
+
+        // First case: the widget already has a captureStream
+        if (typeof this.get('source').captureStream == 'function') {
+            this.captureStream = () => {
+                return this.get('source').captureStream();
+            };
+
+            return;
+        };
+
+        var id_views = Object.keys(this.get('source').views);
+        if (id_views.length == 0) {
+            this.captureStream = () => {
+                return new Promise((resolve, reject) => {
+                    reject({'message': 'Cannot create WidgetStream if the widget has no view rendered'});
+                });
+            };
+
+            return;
+        }
+
+        var first_view = this.get('source').views[id_views[0]];
+        return first_view.then((view) => {
+            // Second case: the widget view is a canvas or a video element
+            var capturable_obj = this.find_capturable_obj(view.el);
+            if (capturable_obj) {
+                this.captureStream = () => {
+                    return new Promise((resolve, reject) => {
+                        if(capturable_obj.captureStream || capturable_obj.mozCaptureStream) {
+                            if(capturable_obj.captureStream) {
+                                resolve(capturable_obj.captureStream());
+                            } else if(capturable_obj.mozCaptureStream) {
+                                resolve(capturable_obj.mozCaptureStream());
+                            }
+                        } else {
+                            reject(new Error('captureStream not supported for this browser'));
+                        }
+                    });
+                };
+
+                return;
+            }
+
+            // Third case: use html2canvas
+            this.canvas = document.createElement('canvas');
+            var updateStream = (el) => {
+                html2canvas(el, {
+                    canvas: this.canvas,
+                    logging: false,
+                    useCORS: true,
+                }).then(() => {
+                    updateStream(el);
+                });
+            };
+
+            this.captureStream = () => {
+                return new Promise((resolve, reject) => {
+                    if(this.canvas.captureStream || this.canvas.mozCaptureStream) {
+                        if(this.canvas.captureStream) {
+                            resolve(this.canvas.captureStream());
+                        } else if(this.canvas.mozCaptureStream) {
+                            resolve(this.canvas.mozCaptureStream());
+                        }
+                    } else {
+                        reject(new Error('captureStream not supported for this browser'));
+                    }
+                });
+            };
+
+            return new Promise((resolve, reject) => {
+                updateStream(view.el);
```

Comment:
We need a method for stopping the infinite recursive call, maybe a `stream` attribute set to true by default, and if it's false we stop the stream with html2canvas.

Reply:
I think we should stop once the widget is closed.

Reply:
When is the widget considered closed? This code is part of the javascript model, so isn't it "alive" as long as the python object exists?

Reply:
Ok thanks :)